### PR TITLE
Popup positioning rework

### PR DESCRIPTION
**Pull request by kekekeks**
Added and implemented IPopupPositioner as described in https://github.com/AvaloniaUI/Avalonia/issues/2405

Aside from being wayland compatible and being decoupled from Popup/PopupImpl logic, the new positioning infrastructure is actually aware of screen bounds and is also aware of placement rectangles when trying to fit the popup into the screen, so the placement target won't be obscured by the popup.

output

Breaking changes:
- Popup now needs either a PlacementTarget or a logical parent attached to the visual tree.
- PopupRoot now always needs a toplevel parent.

### CODE REVIEW

File: src/Avalonia.Controls/Primitives/Popup.cs [hunk 1]
```diff
@@ -218,57 +226,71 @@ public bool Topmost
...
+                    "Attempted to open a popup not attached to a TopLevel");
             }
 
-            _popupRoot.Position = GetPosition();
 
-            if (_topLevel == null && PlacementTarget != null)
+            _popupRoot = new PopupRoot(_topLevel, DependencyResolver)
             {
-                _topLevel = PlacementTarget.GetSelfAndLogicalAncestors().First(x => x is TopLevel) as TopLevel;
+                [~WidthProperty] = this[~WidthProperty],
```
Comment (grokys) [inline src/Avalonia.Controls/Primitives/Popup.cs]: We probably want to rethink binding the popup root's sizing parameters to `Popup`: see https://github.com/AvaloniaUI/Avalonia/issues/2373

Doesn't need to be done in this PR, but thought I'd make you aware of that issue.

File: src/Avalonia.Controls/Primitives/Popup.cs [hunk 2]
```diff
@@ -492,5 +470,54 @@ public void Dispose()
                 _owner._ignoreIsOpenChanged = false;
             }
         }
+
+        // For some reason when PopupRoot.Content is bound directly to Child weird stuff happens
```
Comment (grokys) [inline src/Avalonia.Controls/Primitives/Popup.cs]: Hmm, what weird stuff?

Comment (kekekeks) [reply src/Avalonia.Controls/Primitives/Popup.cs]: Yes, for some reason the content gets stripped of all of its children.

File: src/Avalonia.Controls/Primitives/Popup.cs (hunk 1 above)
Comment (kekekeks) [reply src/Avalonia.Controls/Primitives/Popup.cs]: We probably need to make `Popup` to not be a part of the visual tree. Then we could derive from StyledElement directly and define new Width/Height properties

File: src/Avalonia.Controls/Primitives/Popup.cs (hunk 2 above)
Comment (kekekeks) [reply src/Avalonia.Controls/Primitives/Popup.cs]: I've initially added `PopupContentHost` to debug the visual children disappearance, but it has suddenly stopped to be a problem, so I've left it as is.

Comment (kekekeks) [reply src/Avalonia.Controls/Primitives/Popup.cs]: I think I will need a class like that for overlay popups, so I think we should keep it.

File: tests/Avalonia.Controls.UnitTests/Primitives/PopupTests.cs [hunk 3]
```diff
@@ -233,46 +223,28 @@ public void Popup_Close_Should_Raise_Single_Closed_Event()
...
-            }
-        }
-
+        
         [Fact]
         public void Templated_Control_With_Popup_In_Template_Should_Set_TemplatedParent()
         {
+            if(UsePopupHost)
+                // For some reason with overlay popups templates don't get applied in test mode but
+                // everything works perfectly fine at runtime. I leave this one to you @grokys
```
Comment (kekekeks) [inline tests/Avalonia.Controls.UnitTests/Primitives/PopupTests.cs]: Actually tooltips and DatePicker are broken, so it's not "perfectly fine"

File: src/Avalonia.Controls/Primitives/IPopupHost.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+using System;
+using Avalonia.Controls.Primitives.PopupPositioning;
+using Avalonia.VisualTree;
+
+namespace Avalonia.Controls.Primitives
+{
+    public interface IPopupHost : IDisposable
+    {
+        void SetContent(IControl control);
```
Comment (grokys) [inline src/Avalonia.Controls/Primitives/IPopupHost.cs]: `Content` usually refers to data that can be anything, e.g. `object Content`. If the content is a control, the property is usually named `Child`.

File: src/Avalonia.Controls/Primitives/IPopupHost.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        IVisual VisualRoot { get; }
```
Comment (grokys) [inline src/Avalonia.Controls/Primitives/IPopupHost.cs]: Why is this property needed? It returns `this` in both implementations of this class.

File: src/Avalonia.Controls/Primitives/OverlayLayer.cs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            InvalidateMeasure();
+            InvalidateArrange();
+        }
+
+        public void Remove(Control v) => VisualChildren.Remove(v);
+        
+        protected override Size MeasureOverride(Size availableSize)
+        {
+            
+            var infinite = new Size(double.PositiveInfinity, double.PositiveInfinity);
```
Comment (grokys) [inline src/Avalonia.Controls/Primitives/OverlayLayer.cs]: There's already `Size.Infinite`.

File: src/Avalonia.Controls/Primitives/IPopupHost.cs (hunk 5 above)
Comment (kekekeks) [reply src/Avalonia.Controls/Primitives/IPopupHost.cs]: It returns `null` for `PopupHost`. It's needed for devtools, I believe.

File: src/Avalonia.Controls/Primitives/VisualLayerManager.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+namespace Avalonia.Controls.Primitives
+{
+    public class VisualLayerManager : Decorator
+    {
+        private const int AdornerZIndex = int.MaxValue - 100;
+        private const int OverlayZIndex = int.MaxValue - 99;
+
+        private bool _isAttachedToLogicalTree;
+        private IStyleHost _styleHost;
```
Comment (grokys) [inline src/Avalonia.Controls/Primitives/VisualLayerManager.cs]: Nit: I think this should be called `_styleRoot`.

File: src/Avalonia.Controls/Primitives/VisualLayerManager.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        public bool IsPopup { get; set; }
```
Comment (grokys) [inline src/Avalonia.Controls/Primitives/VisualLayerManager.cs]: Nit: Can we move this to be with other properties? It gets mistaken as a private field here.

File: src/Avalonia.Controls/Primitives/VisualLayerManager.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        
+        List<Control> _layers = new List<Control>();
```
Comment (grokys) [inline src/Avalonia.Controls/Primitives/VisualLayerManager.cs]: Nit: missing `private readonly`.

File: src/Avalonia.Controls/Primitives/PopupPositioning/ManagedPopupPositioner.cs [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+            Update(_popup.TranslateSize(parameters.Size), parameters.Size,
+                new Rect(_popup.TranslatePoint(parameters.AnchorRectangle.TopLeft),
+                    _popup.TranslateSize(parameters.AnchorRectangle.Size)),
+                parameters.Anchor, parameters.Gravity, parameters.ConstraintAdjustment,
+                _popup.TranslatePoint(parameters.Offset));
+        }
+
+        
+        void Update(Size translatedSize, Size originalSize,
```
Comment (grokys) [inline src/Avalonia.Controls/Primitives/PopupPositioning/ManagedPopupPositioner.cs]: Nit: missing `private`.

File: src/Avalonia.Themes.Default/ComboBox.xaml [hunk 11]
```diff
@@ -39,7 +39,7 @@
                    StaysOpen="False">
               <Border BorderBrush="{DynamicResource ThemeBorderMidBrush}"
                       BorderThickness="1">
-                  <AdornerDecorator Margin="-1 -1 0 0">
+                  <VisualLayerManager IsPopup="True" Margin="-1 -1 0 0">
```
Comment (grokys) [inline src/Avalonia.Themes.Default/ComboBox.xaml]: Now that `PopupHost` is a templated control too, we can probably move this into the template?

File: src/Avalonia.Visuals/Media/PixelPoint.cs [hunk 12]
```diff
@@ -106,6 +159,8 @@ public override int GetHashCode()
                 return hash;
             }
         }
+        
+
```
Comment (grokys) [inline src/Avalonia.Visuals/Media/PixelPoint.cs]: Nit: whitespace changes.

File: src/Avalonia.Visuals/Media/PixelRect.cs [hunk 13]
```diff
@@ -261,6 +261,17 @@ public bool Intersects(PixelRect rect)
...
+        /// <summary>
+        /// Translates the rectangle by an offset.
+        /// </summary>
+        /// <param name="offset">The offset.</param>
+        /// <returns>The translated rectangle.</returns>
+        public PixelRect Translate(PixelVector offset)
+        {
+            return new PixelRect(Position + offset, Size);
+        }
+
```
Comment (grokys) [inline src/Avalonia.Visuals/Media/PixelRect.cs]: Nit: whitespace.

File: src/Avalonia.Visuals/Media/PixelVector.cs [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+        }
+
+        /// <summary>
+        /// Check if two vectors are equal (bitwise).
+        /// </summary>
+        /// <param name="other"></param>
+        /// <returns></returns>
+        public bool Equals(PixelVector other)
+        {
+            // ReSharper disable CompareOfFloatsByEqualityOperator
```
Comment (grokys) [inline src/Avalonia.Visuals/Media/PixelVector.cs]: I think we can remove these resharper pragmas as we're not comparing floats here.

**Review by grokys — CHANGES_REQUESTED** [review]

**Pull request merged**